File: lua/day16_part2_2021.lua

local function hexToBin(hex)
    local bin = {}
    for i = 1, #hex do
        local b = tonumber(hex:sub(i,i), 16)
        local bits = {}
        for j = 3, 0, -1 do
            bits[#bits+1] = math.floor(b / (2^j)) % 2
        end
        for _, bit in ipairs(bits) do
            bin[#bin+1] = tostring(bit)
        end
    end
    return table.concat(bin)
end

local function parsePacket(binStr, idx)
    local function bitsToNum(start, length)
        local num = 0
        for i = start, start + length - 1 do
            num = num * 2 + tonumber(binStr:sub(i,i))
        end
        return num, start + length
    end

    local version, newIdx = bitsToNum(idx, 3)
    local typeID, newIdx = bitsToNum(newIdx, 3)

    if typeID == 4 then
        local value = 0
        while true do
            local group = binStr:sub(newIdx, newIdx)
            local num, nextIdx = bitsToNum(newIdx + 1, 4)
            value = value * 16 + num
            newIdx = nextIdx
            if group == '0' then break end
        end
        return version, newIdx, value
    end

    local lengthTypeID = tonumber(binStr:sub(newIdx, newIdx))
    newIdx = newIdx + 1

    local values = {}
    local subPacketLength, numSubPackets = 0, 0

    if lengthTypeID == 0 then
        subPacketLength, newIdx = bitsToNum(newIdx, 15)
    else
        numSubPackets, newIdx = bitsToNum(newIdx, 11)
    end

    local startIdx = newIdx
    while true do
        if (lengthTypeID == 0 and subPacketLength == 0) or 
           (lengthTypeID == 1 and numSubPackets == 0) then
            break
        end

        local _, nextIdx, subValue = parsePacket(binStr, newIdx)
        table.insert(values, subValue)

        if lengthTypeID == 0 then
            subPacketLength = subPacketLength - (nextIdx - newIdx)
        else
            numSubPackets = numSubPackets - 1
        end
        newIdx = nextIdx
    end

    local result
    if typeID == 0 then
        result = 0
        for _, v in ipairs(values) do result = result + v end
    elseif typeID == 1 then
        result = 1
        for _, v in ipairs(values) do result = result * v end
    elseif typeID == 2 then
        result = math.min(table.unpack(values))
    elseif typeID == 3 then
        result = math.max(table.unpack(values))
    elseif typeID == 5 then
        result = values[1] > values[2] and 1 or 0
    elseif typeID == 6 then
        result = values[1] < values[2] and 1 or 0
    elseif typeID == 7 then
        result = values[1] == values[2] and 1 or 0
    else
        error("Unknown typeID")
    end

    return version, newIdx, result
end

local file = io.open("input.txt", "r")
local hexStr = file:read("*all"):gsub("%s", "")
file:close()

local binStr = hexToBin(hexStr)
local _, _, value = parsePacket(binStr, 1)
print(value)

File: julia/day16_part2_2021.jl

function hex2bin(hex)
    bin = ""
    for h in hex
        b = parse(UInt8, h, base=16)
        bin *= lpad(string(b, base=2), 4, '0')
    end
    return bin
end

function parse_packet(bin_str, idx)
    version = parse(Int, bin_str[idx:idx+2], base=2)
    type_id = parse(Int, bin_str[idx+3:idx+5], base=2)
    idx += 6

    if type_id == 4
        value = 0
        while bin_str[idx] == '1'
            value = value << 4 + parse(Int, bin_str[idx+1:idx+4], base=2)
            idx += 5
        end
        value = value << 4 + parse(Int, bin_str[idx+1:idx+4], base=2)
        idx += 5
        return version, idx, value
    end

    length_type_id = parse(Int, bin_str[idx])
    idx += 1
    
    if length_type_id == 0
        sub_packet_length = parse(Int, bin_str[idx:idx+14], base=2)
        idx += 15
    else
        num_sub_packets = parse(Int, bin_str[idx:idx+10], base=2)
        idx += 11
    end

    values = Int64[]
    if length_type_id == 0
        target_idx = idx + sub_packet_length
        while idx < target_idx
            _, new_idx, sub_value = parse_packet(bin_str, idx)
            push!(values, sub_value)
            idx = new_idx
        end
    else
        for _ = 1:num_sub_packets
            _, new_idx, sub_value = parse_packet(bin_str, idx)
            push!(values, sub_value)
            idx = new_idx
        end
    end

    if type_id == 0
        result = sum(values)
    elseif type_id == 1
        result = prod(values)
    elseif type_id == 2
        result = minimum(values)
    elseif type_id == 3
        result = maximum(values)
    elseif type_id == 5
        result = values[1] > values[2] ? 1 : 0
    elseif type_id == 6
        result = values[1] < values[2] ? 1 : 0
    elseif type_id == 7
        result = values[1] == values[2] ? 1 : 0
    else
        error("Unknown typeID")
    end

    return version, idx, result
end

data = read("input.txt", String)
hex_str = strip(data)
bin_str = hex2bin(hex_str)
_, _, value = parse_packet(bin_str, 1)
println(value)
